Sources/ClientRuntime/Networking/Streaming/BufferedStream.swift

```diff
@@ -0,0 +1,128 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// A `Stream` implementation that buffers data in memory.
+/// The buffer size depends on the amount of data written and read.
+/// Note: This class is thread-safe.
+/// Note: if data is not read from the stream, the buffer will grow indefinitely until the stream is closed.
+///       or reach the maximum size of a `Data` object.
+public class BufferedStream: Stream {
+
+    /// Returns the length of the stream, if known
+    public private(set) var length: Int?
+
+    /// Returns the current position in the stream
+    public private(set) var position: Data.Index
+
+    /// Returns true if the stream is empty, false otherwise
+    public var isEmpty: Bool {
+        return buffer.isEmpty
+    }
+
+    /// Returns true if the stream is seekable, false otherwise
+    /// Note: This stream is seekable, but seeking to a position that is not in the buffer will cause the stream to throw an error.
+    public var isSeekable: Bool {
+        return false
+    }
```

Comment:
Does this value ever change or is it static?

Reply:
And should this value be `true` since we have defined an implementation for seeking?

Reply:
here is bit controversy, you can only seek in the buffer but this property tells that whole stream is seekable.

Reply:
Its also a bit odd to have this value return false yet we define an implementation for the public seek method. 
Something isn't quite right here.

Reply:
made some changes around this, no longer applicable.